README
WiControl
=========

Wireless Control with Panstamp and Raspberry Pi

The libserial-0.6.0rc1 directory contains a version of the C++ libserial
library configured for the Raspberry Pi.

Installation:
-------------

On the Pi:

$ mkdir Panstamp

$ cd Panstamp

$ git clone https://github.com/AdrianRepRap/WiControl.git

$ cd WiControl/libserial-0.6.0rc1

$ make clean

$ sudo make install

$ cd ../wput-0.6.1

$ ./configure

$ make

$ sudo make install

$ cd ../Host-software/Debug

$ make clean

$ make

$ cd ../..

$ cp run-heat-control.sample run-heat-control

then edit run-heat-control to download your control file from your webserver and to 
run the program with that file.  Replace the ***** with your ftp password so that the 
temperatures can be uploaded to the server.  (Remember after that to keep the file
private - in particular don't push it to Github...)  Finally

$ sudo crontab -e

and add lines at the end like this:

* * * * * /home/pi/Panstamp/WiControl/run-heat-control

0 4   *   *   *    /sbin/shutdown -r +5

That will cause the shell script run-heat-control to be run once a minute, and cause a
system reboot at 04:05 in the morning to keep everything clean and tidy.  The sudo is
needed because shutdown needs root privilege.

When it runs it creates a status file called Status.

$ cat Status

will tell you what your system last did.  If you

$ touch Log

then each time run-heat-control runs it will append the status
report to the end of itself, which keeps a history and is therefore
useful for debugging.

When you have finished debugging don't forget to

$ rm Log

otherwise it will eventually fill up the SD/disc...


Adrian Bowyer



wput-0.6.1/ChangeLog
Wed Dec 11 20:53:52 CET 2007 wput-0.6.1
	* adding russian (ru) translation by Vyacheslav Nikitin
	* building now to wput-VERSION fixing #1844014

Sun Dec 02 16:55:29 CET 2007
	* fixing #1776300 (build issues), #1834977 (netrc),
	  #1737168 (return values), #1651042 (resume-corruption),
	  #1599242 (TYPE<->SIZE), #1525208 (ps and passwords),
	  #1523072 (timeoffset)
	* adding simplified chinese (zh) translation by Sender Soff
	* adding turkish (tr) translation by Eren Türkay

Wed Apr 12 18:17:32 CET 2006
	* adding dutch (nl) translation by Sebastiaan van der Velden

Sat Apr 08 15:11:50 CET 2006
	* gnutls created SIGPIPE events when the connection broke
	  down. ignoring SIGPIPE now
	* fixing some issues in transferrate/eta-calculation for
	  slower rates than 10K

Fri Apr 07 14:12:33 CET 2006
	* fixed the bar_update function which was too busy printing
	  out updates. limited to 10 updates per second
	* fixed the bar_creating function which called get_terminal_size
	  even in background mode, causing strange errors afterwards

Thu Apr 06 23:05:49 CET 2006
	* changed SOCKET_ERROR to SOCK_ERROR because SOCKET_ERROR
	  is somewhat defined by winsock in w32
	* minor other changes to help compilation for win32

Wed Apr 05 12:57:10 CET 2006
	* fixing a bug in get_relative_path which returned
	  false data for pathcombinations like "a","abc"

Tue Apr 04 18:46:34 CET 2006 wput-0.6
	* using gnutls now. adjusting includes. no need for win32
	* parsing of mdtm forgot to decrease the month. fixed
	* setting target_fsize to -1 to indicate unexisting files,
	  so that wput will upload even files with 0 filesize
	* fixing double free() of hostname if the host could not be
	  resolved
	* fixing reported bug in Makefile
	* adding support for .netrc-files as suggested by a patch
	  of Alexander Pohoyda
	* when sending or parsing PORT/PASV commands, now using
	  htons to convert to network byte order

Thu May 26 09:30:54 CET 2005
	* buffer-allocation in http-proxy-module forgot to count five
	  additional bytes for the port-value which led to segfaults

Sun May  1 10:00:16 CET 2005
	* fixed a calculation for calculate_transfer_rate that caused an
	  data-overflow for rates beyond 210 KiB/s
	* another fix for set_resuming because the last fix made things even
	  worse. works fine now ;)

Fri Apr 29 16:53:14 CET 2005
	* minor fix in set_resuming which mixed up some cases

Wed Mar 23 10:32:41 CET 2005
	* fix for base64-routine which forgot to set the last character to \0

Sat Feb  5 20:37:09 CET 2005 wput-pre0.6
	* making compilation for windows work (ssllib in dll, some
	  modifications (sslib_in_use flag), strptime included, some
	  fixes for *nix-header-files)

Tue Feb  2 22:10:45 CET 2005
	* added support for i18n (configure, gettext, setlocale etc.)
	* messages have been prepared
	* german translation is complete

Sun Jan 30 14:00:56 CET 2005 wput-0.5.9
	* added the --no-directories flag forcing wput not to create any
	  remote directories.
	* added the --basename param that allows to specify a base-path-
	  name which is snipped off the filename when connecting them to
	  an url

Sat Jan 29 13:54:26 CET 2005
	* set up an error-level-indication (see manpage)
	* added additional status-information telling about the number of
	  files failed and skipped
	* made LIST work again
	* directory_add_dir() needs to save a cpy of the directory name

Thu Jan 27 19:23:01 CET 2005
	* [ lots of changes. dull brain had a memory-leak and forgot them ]
	* fixed some issues with resume-tables
	* fixed the get_relative_path function
	* removed several memory-allocation problems with the new ftp/socket-
	  libary objects

Sun Jan 23 11:27:30 CET 2005
	* created an ftp-library ftplib.[ch] which contains more or less
	  the general ftp-function from ftp.c. Those functions work on an
	  ftp_con-object which is independent of any fsession. Each
	  fession contains a reference to an ftp_con-object representing
	  the current connection. The cc-object therefore is gone.
	* do_send was cleaned up and lots of its code was moved to some
	  seperate functions
	* fsession_transmit moved to ftp.c with some minor modifications
	* added a function proxy_accept which is used by
	  ftp_complete_data-connection which does the piece of work needed
	  to complete a data-connection after a STOR or LIST command

Thu Jan 20 11:22:13 CET 2005
	* renamed socket.[ch] to socketlib.[ch] as these reserved names
	  caused some trouble on some systems
	* code-cleanup, more general functions and a slightly more object-
	  orientated design bring socketlib quit near to a standalone module
	* all the constant-defines moved from wput.h to constants.h

Wed Jan 19 09:17:02 CET 2005
	* basic implementation of tls-encryption
	* adjustment of configure.in and config.h.in to determine whether
	  openssl is available
	* config.h moved to src/
	* ftp-ls.c is now wget's one made workable with some
	  #define directives and some dummy-functions as proposed by
	  Steven M. Schweda
	* fsessions are automatically sorted into the queue

Sun Jan 09 22:44:19 CET 2005
	* parse_url now looks for the last @ char, thus usernames containing
	  an @ will be valid as well

Mon Nov 22 20:32:07 CET 2004	wput 0.5
	* clear_path was also called when target_dname was 0 > segfault. fixed
	* queue.c#409 removes not only strlen(file) chars, but strlen(file)-1
	* get_msg() sets cc.r.reply to 0 after freeing it

Fri Nov 19 22:42:43 CET 2004
	* little fixes for problems causing troubles when compiling for win32
	* fix for do_mdtm which forgot to parse the returned date
	* do_abor() and functions using it, return ERR_RECONNECT if it fails
	* updated the documentation (manpage,USAGE.examples: timestamping)

Thu Nov 18 13:45:57 CET 2004
	* queue.c#469 now creates a cpy of fsession->local_fname (otherwise
	  free()ing causes undefined behaviour)...
	* fixed the problem in bar_update() that caused the old-style
	  interface to output 101K, 151K and so on instead of n*50K...

Control/Control.h
// Only modify this file to include
// - function definitions (prototypes)
// - include files
// - extern variable definitions
// In the appropriate section

#ifndef Control_H_
#define Control_H_
#include "Arduino.h"
#include <avr/wdt.h>

#include "Communicator.h"
#include "CommandBuffer.h"
#include "Switch.h"

#define MY_ADDRESS 10

#define BAUD_RATE 38400

#define NAME "WiControl"
#define DATE "2014-08-24"
#define VERSION "0.12"

// Comment this out to turn off the watchdog reset
#define WATCHDOG

// Uncomment this to use LED control
//#define USE_LEDS

#define DATA_LENGTH 60

// The LED is wired to the Arduino Output 4 (physical panStamp pin 19)
#define LEDOUTPUT 4

#ifdef USE_LEDS
// RGB PWM outputs

#define RED_PWM 6
#define GREEN_PWM 5
#define BLUE_PWM 9
#endif

#define RESET_SECONDS 90

#define ABS_ZERO -273.15  // Celsius
#define TEMP_SENSE_PIN 0   // Analogue pin number

// Outputs for controlling various things.  D_OUTPUT_1 is the MOSFET
// D_OUTPUT_2 is the first output on the 7 pin header, which from the top goes:
// 12v, Gnd, D8, D9, A3 = D17, A1 = D15, A2 = D16
// D_OUTPUT_3 is not available if LEDS are being controlled

#ifndef USE_LEDS
 #define SWITCHES 7
 #define OUT_PINS {3, 8, 9, 17, 15, 16, LEDOUTPUT}
#else
 #define SWITCHES 6
 #define OUT_PINS {3, 8, 17, 15, 16, LEDOUTPUT};
#endif

#define THERMISTOR_BETA 3528.0 // thermistor: RS 538-0806
#define THERMISTOR_SERIES_R 1000 // Ohms in series with the thermistors
#define THERMISTOR_25_R 1000.0 // Thermistor ohms at 25 C = 298.15 K
#define AD_RANGE 1023.0 // The A->D converter that measures temperatures gives an int this big as its max value


void IncommingInterrupt();
void blink();

#ifdef __cplusplus
extern "C" {
#endif
void loop();
void setup();
#ifdef __cplusplus
} // extern "C"
#endif

#endif /* Control_H_ */
